vm/devices/firmware/firmware_uefi/src/service/diagnostics.rs

```diff
@@ -0,0 +1,476 @@
+        let mut phase = 0;
+        let mut is_accumulating = false;
+
+        // Used for tracking what has been processed
+        let mut bytes_read: usize = 0;
+        let mut entries_processed: usize = 0;
+
+        // Process the buffer slice until all entries are processed
+        while !buffer_slice.is_empty() {
+            // Parse the current etnry
+            let entry = parse_entry(buffer_slice)?;
+
+            // Handle message accumulation
+            if !is_accumulating {
+                debug_level = entry.debug_level;
+                time_stamp = entry.time_stamp;
+                phase = entry.phase;
+                accumulated_message.clear();
+                is_accumulating = true;
+            }
+
+            // Append the message
+            accumulated_message.push_str(entry.message);
+
+            // Check message length
+            if accumulated_message.len() > MAX_MESSAGE_LENGTH as usize {
+                return Err(DiagnosticsError::BadAccumulatedMessageLength(
+                    MAX_MESSAGE_LENGTH,
+                    accumulated_message.len() as u16,
+                ));
+            }
+
+            // Handle completed messages (ending with '\n')
+            if !entry.message.is_empty() && entry.message.ends_with('\n') {
+                log_handler(EfiDiagnosticsLog {
+                    debug_level,
+                    ticks: time_stamp,
+                    phase,
+                    message: std::mem::take(&mut accumulated_message)
+                        .trim_end_matches(&['\r', '\n'][..])
+                        .to_string(),
+                });
+                entries_processed += 1;
+                is_accumulating = false;
+            }
+
+            // Update bytes read and move to the next entry
+            bytes_read = bytes_read.checked_add(entry.entry_size).ok_or_else(|| {
+                DiagnosticsError::Overflow(anyhow!(
+                    "bytes_read ({}) + entry_size ({})",
+                    bytes_read,
+                    entry.entry_size
+                ))
+            })?;
+
+            // Advance the buffer to the next entry
+            if entry.entry_size >= buffer_slice.len() {
+                break; // End of buffer
+            } else {
+                buffer_slice = &buffer_slice[entry.entry_size..];
+            }
+        }
+
+        // Process any remaining accumulated message
+        if is_accumulating && !accumulated_message.is_empty() {
+            log_handler(EfiDiagnosticsLog {
+                debug_level,
+                ticks: time_stamp,
+                phase,
+                message: std::mem::take(&mut accumulated_message)
+                    .trim_end_matches(&['\r', '\n'][..])
+                    .to_string(),
+            });
+            entries_processed += 1;
+        }
+
+        // Print summary statistics
+        tracelimit::info_ratelimited!(
+            "EFI Diagnostics: Processed {} entries, Read {} bytes",
```

Comment:
Do not use string interpolation for trace statements. Use key value pairs, always.

`tracelimit::info_ratelimited!(entries_processes, bytes_read, "processed EFI log entries");`